src/cargo/core/resolver/mod.rs

```diff
@@ -127,15 +129,22 @@ pub fn resolve(
     config: Option<&Config>,
     check_public_visible_dependencies: bool,
 ) -> CargoResult<Resolve> {
-    let cx = Context::new(check_public_visible_dependencies);
     let _p = profile::start("resolving");
     let minimal_versions = match config {
         Some(config) => config.cli_unstable().minimal_versions,
         None => false,
     };
-    let mut registry =
-        RegistryQueryer::new(registry, replacements, try_to_use, minimal_versions, config);
-    let cx = activate_deps_loop(cx, &mut registry, summaries, config)?;
+    let (registry, cx) = loop {
+        let mut registry =
+            RegistryQueryer::new(registry, replacements, try_to_use, minimal_versions, config);
+        let cx = Context::new(check_public_visible_dependencies);
+        let cx = activate_deps_loop(cx, &mut registry, summaries, config)?;
+        if registry.all_ready() {
+            break (registry, cx);
+        } else {
+            // TODO: dont hot loop for it to be Ready
```

Comment:
FWIW I'm imagining a new method on `Registry` which is something like "block on things" which is called here.